PINCache/PINOperationQueue.m

```diff
@@ -0,0 +1,217 @@
+    pthread_mutex_init(&_lock, &attr);
+    
+    _serialQueue = dispatch_queue_create("PINOperationQueue Serial Queue", DISPATCH_QUEUE_SERIAL);
+    
+    _concurrentQueue = dispatch_queue_create("PINOperationQueue Unprioritized Serial Queue", DISPATCH_QUEUE_CONCURRENT);
+    _concurrentSemaphore = dispatch_semaphore_create(maxConcurrentOperations - 1);
+    _semaphoreQueue = dispatch_queue_create("PINOperationQueue Serial Semaphore Queue", DISPATCH_QUEUE_SERIAL);
+    
+    _allOperations = [[NSMutableOrderedSet alloc] init];
+    _lowPriorityOperations = [[NSMutableOrderedSet alloc] init];
+    _defaultPriorityOperations = [[NSMutableOrderedSet alloc] init];
+    _highPriorityOperations = [[NSMutableOrderedSet alloc] init];
+  }
+  return self;
+}
+
+- (void)dealloc
+{
+  pthread_mutex_destroy(&_lock);
+}
+
+- (id <PINOperationReference>)nextOperationReference
+{
+  [self lock];
+    id <PINOperationReference> reference = [NSNumber numberWithUnsignedInteger:++_operationReferenceCount];
+  [self unlock];
+  return reference;
+}
+
+- (id <PINOperationReference>)addOperation:(dispatch_block_t)block withPriority:(PINOperationQueuePriority)priority
+{
+  id <PINOperationReference> reference = [self nextOperationReference];
+  
+  NSMutableOrderedSet *queue;
+  switch (priority) {
+    case PINOperationQueuePriorityLow:
+      queue = _lowPriorityOperations;
+      break;
+      
+    case PINOperationQueuePriorityDefault:
+      queue = _defaultPriorityOperations;
+      break;
+      
+    case PINOperationQueuePriorityHigh:
+      queue = _highPriorityOperations;
+  }
+  
+  PINOperation *operation = [PINOperation operationWithBlock:block reference:reference];
+  
+  [self lock];
+    [queue addObject:operation];
+    [_allOperations addObject:operation];
+  [self unlock];
+  
+  [self scheduleOperations:NO];
+  
+  return reference;
+}
+
+- (void)cancelOperation:(id <PINOperationReference>)operationReference
+{
+  [self lock];
+    [_canceledOperations addObject:operationReference];
+  [self unlock];
+}
+
+- (void)scheduleOperations:(BOOL)onlyCheckSerial
+{
+  [self lock];
+    //get next available operation in order, ignoring priority and run it on the serial queue
+    if (_serialQueueBusy == NO) {
+      PINOperation *operation = [_allOperations firstObject];
+      if (operation) {
+        _serialQueueBusy = YES;
+        [self removeOperation:operation];
+        dispatch_async(_serialQueue, ^{
+          operation.block();
+          [self lock];
+            _serialQueueBusy = NO;
+          [self unlock];
+          
+          //see if there are any other operations
+          [self scheduleOperations:YES];
+        });
+      }
+    }
+  [self unlock];
+  
+  if (onlyCheckSerial) {
+    return;
+  }
+  
+  dispatch_async(_semaphoreQueue, ^{
+      dispatch_semaphore_wait(_concurrentSemaphore, DISPATCH_TIME_FOREVER);
+      [self lock];
+        PINOperation *operation = [self nextOperationByPriority];
```

Comment:
I'd expect a method called `scheduleOperations:` to schedule all operations. Unless I'm wrong, you schedule only the top priority one here.
It might bite you later if you decide to add a `addOperations:withPriority:` method because you'd reasonably call `scheduleOperations:` once.